system-configuration/src/network_reachability.rs

```diff
@@ -434,6 +453,19 @@ mod test {
         }
     }
 
+    #[test]
+    fn test_sockaddr_local_to_dns_google_pair_reachability() {
+        let Ok(dns_google) = std::net::TcpStream::connect("8.8.4.4:443") else {
+            eprintln!("test skipped due to inability to connect to well-known endpoint (dns.google:443)");
+            return;
+        };
+        let local = dns_google.local_addr().unwrap();
+        let remote = dns_google.peer_addr().unwrap();
```

Comment:
Why do you need to create an actual connection to just get a set of `SocketAddr` instances? This makes the test need external dependencies and flakier. Why not just create these two, or any set of addresses with `SocketAddr::new` or similar?

Reply:
I wanted to test actual reachability of socket pairs, even so I used well known remote address, I don't known local one to create reachable pair. The `SCNetworkReachabilityCreateWithAddressPair` API is a must for bsd sockets reachability monitoring [according to apple](https://developer.apple.com/documentation/technotes/tn3151-choosing-the-right-networking-api#BSD-Sockets-best-practices):
> Once you’ve established a connection, use [SCNetworkReachabilityCreateWithAddressPair(_:_:_:)](https://developer.apple.com/documentation/systemconfiguration/1514908-scnetworkreachabilitycreatewitha) to monitor its viability. Without this, your program won’t notice that a connection is stuck due to a TCP/IP stack reconfiguration.

I've also reworked test and now try to connect multiple well known hosts (I assume at least one of them is reachable from test runner) and verify that in case of connection error reachability is reported as not reachable, and in case of successful connection it is reported as reachable. 

It still flaky in a sense that inability to connect to well known host might happen in case of network route exist (aka socket pair is reachable), but I think it will happen very rarely, while tests still checks valuable behaviour of library.

Reply:
This test then has external dependencies and becomes flaky. If run in an environment without internet, it will always just silently work, without ever testing anything. It will say that the function works, without testing it. And even in an environment with internet it becomes flaky.

So it's not a unit test any longer, but more of an integration test. Would be nice if that was differentiated so that just `cargo test` does not come with implicit external dependencies. This is a common problem and I don't know of a de facto standard of solving it. But putting these integration tests behind some `cfg` flag or similar could work. But it would be nice if the output then also reflected that some tests were ignored.

I'll think a bit about it. But I have bad experience with just assuming things will work most of the time. Given enough tests and complexity, the amount of times *something* will fail becomes way too high and people stop respecting the test matrix.

Reply:
>  it will always just silently work, without ever testing anything. It will say that the function works, without testing 

I've modified test in last commit to verify that in case of inability to connect (like no internet case) it will be checked that reachability also report unreachable status, so it's not completely useless without internet. The only flakiness left is dns.google not accepting tcp connections for some reason or firewall blocks them - so the reachability will report reachable in such case, while actual connection is failed. 

> But putting these integration tests behind some cfg flag or similar could work. But it would be nice if the output then also reflected that some tests were ignored.

Maybe it can be moved to integration tests folder `/tests` so it will be clear (by convention) that test is integration test.

Or it could be removed or replaced with different test which only checks the correctness of `to_c_sockaddr`.

Reply:
we can maybe merge it as is, and only improve if it becomes an actual issue that it's flaky! Let me try this out and see.

My understanding is that the actual bug here is that both the port and the ip fields, when translated to the C versions, were not properly converted to big endian, right?

Reply:
yes, it only worked correctly for `0.0.0.0:0`-like addresses, when I tried to use API to follow reachability of connected socket - it was reported immediately as not reachable, both because local (like 192.168.23.23) and remote address (8.8.4.4) were converted incorrectly and hence were reported as unreachable.